`IndexOutOfBoundsException` when using a view and ordering by a column that isn't selected

### CrateDB version

5.10.0 / 5.9.9

### CrateDB setup information

Standard node

### Problem description

There is an exception for the query given below.

### Steps to Reproduce

```sql
CREATE TABLE IF NOT EXISTS "doc"."t1" (
   "id" INTEGER NOT NULL,
   "propertyName" TEXT NOT NULL,
   "valueDatetime" TIMESTAMP WITHOUT TIME ZONE
);

CREATE TABLE IF NOT EXISTS "doc"."t2" (
   "id" INTEGER NOT NULL,
   "propertyName" TEXT NOT NULL,
   "valueDatetime" TIMESTAMP WITHOUT TIME ZONE
);

INSERT INTO "doc"."t1" VALUES (1, 'test', NOW());
INSERT INTO "doc"."t2" VALUES (1, 'test', NOW());

CREATE OR REPLACE VIEW "doc"."v1" AS
SELECT "id", "propertyName", "valueDatetime"
FROM "doc"."t1"
UNION ALL
SELECT "id", "propertyName", "valueDatetime"
FROM "doc"."t2";

SELECT "id"
FROM "doc"."v1" v
ORDER BY v."valueDatetime"
LIMIT 100;
```

### Actual Result

```
io.crate.exceptions.SQLParseException: Couldn't create execution plan from logical plan because of: Index 1 out of bounds for length 1:
Rename[id] AS v (rows=unknown)
  └ Rename[id] AS doc.v1 (rows=unknown)
    └ Limit[100::bigint;0] (rows=unknown)
      └ Union[id] (rows=unknown)
        ├ OrderBy["valueDatetime" ASC] (rows=unknown)
        │  └ Collect[doc.t1 | [id, "valueDatetime"] | true] (rows=unknown)
        └ OrderBy["valueDatetime" ASC] (rows=unknown)
          └ Collect[doc.t2 | [id, "valueDatetime"] | true] (rows=unknown)
	at io.crate.exceptions.SQLExceptions.esToCrateException(SQLExceptions.java:211)
	at io.crate.exceptions.SQLExceptions.prepareForClientTransmission(SQLExceptions.java:200)
	at io.crate.rest.action.SqlHttpHandler.sendResponse(SqlHttpHandler.java:176)
	at io.crate.rest.action.SqlHttpHandler.lambda$channelRead0$0(SqlHttpHandler.java:130)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:907)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:931)
	at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2401)
	at io.crate.rest.action.SqlHttpHandler.channelRead0(SqlHttpHandler.java:128)
	at io.crate.rest.action.SqlHttpHandler.channelRead0(SqlHttpHandler.java:87)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:107)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.crate.protocols.http.HttpBlobHandler.channelRead0(HttpBlobHandler.java:159)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.crate.auth.HttpAuthUpstreamHandler.handleHttpChunk(HttpAuthUpstreamHandler.java:158)
	at io.crate.auth.HttpAuthUpstreamHandler.channelRead0(HttpAuthUpstreamHandler.java:95)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:107)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:289)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at org.elasticsearch.transport.netty4.Netty4InboundStatsHandler.channelRead(Netty4InboundStatsHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1503)
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1366)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1415)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:530)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:469)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Thread.java:1575)
Caused by: java.lang.IllegalArgumentException: Couldn't create execution plan from logical plan because of: Index 1 out of bounds for length 1:
Rename[id] AS v (rows=unknown)
  └ Rename[id] AS doc.v1 (rows=unknown)
    └ Limit[100::bigint;0] (rows=unknown)
      └ Union[id] (rows=unknown)
        ├ OrderBy["valueDatetime" ASC] (rows=unknown)
        │  └ Collect[doc.t1 | [id, "valueDatetime"] | true] (rows=unknown)
        └ OrderBy["valueDatetime" ASC] (rows=unknown)
          └ Collect[doc.t2 | [id, "valueDatetime"] | true] (rows=unknown)
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302)
	at java.base/java.util.Objects.checkIndex(Objects.java:365)
	at java.base/java.util.ArrayList.get(ArrayList.java:428)
	at io.crate.common.collections.Lists$LazyMapList.get(Lists.java:385)
	at io.crate.execution.dsl.projection.EvalProjection.castValues(EvalProjection.java:58)
	at io.crate.planner.operators.Union.addCastsForIncompatibleObjects(Union.java:171)
	at io.crate.planner.operators.Union.build(Union.java:100)
	at io.crate.planner.operators.Limit.build(Limit.java:111)
	at io.crate.planner.operators.Rename.build(Rename.java:193)
	at io.crate.planner.operators.Rename.build(Rename.java:193)
	at io.crate.planner.operators.RootRelationBoundary.build(RootRelationBoundary.java:60)
	at io.crate.planner.operators.LogicalPlanner.getNodeOperationTree(LogicalPlanner.java:620)
	at io.crate.planner.operators.LogicalPlanner.doExecute(LogicalPlanner.java:599)
	at io.crate.planner.operators.LogicalPlanner.execute(LogicalPlanner.java:569)
	at io.crate.planner.operators.LogicalPlan.executeOrFail(LogicalPlan.java:216)
	at io.crate.planner.Plan.execute(Plan.java:74)
	at io.crate.session.Session.singleExec(Session.java:832)
	at io.crate.session.Session.execute(Session.java:583)
	at io.crate.rest.action.SqlHttpHandler.executeSimpleRequest(SqlHttpHandler.java:273)
	at io.crate.rest.action.SqlHttpHandler.handleSQLRequest(SqlHttpHandler.java:220)
	at io.crate.rest.action.SqlHttpHandler.channelRead0(SqlHttpHandler.java:127)
	... 59 more
Caused by: java.lang.IndexOutOfBoundsException: Index 1 out of bounds for length 1
	... 82 more
```

### Expected Result

Successful execution.

--- a/docs/appendices/release-notes/5.10.1.rst
+++ b/docs/appendices/release-notes/5.10.1.rst
@@ -96,3 +96,14 @@ Fixes
     (SELECT y FROM unnest([2]) as t2(y))
     FROM tbl t
 
+- Fixed an issue that caused a planner exception when using a query containing
+  an ``ORDER BY`` clause on top of a ``UNION`` clause, while the ordering column
+  is not part of the most-top select items. Example::
+
+    SELECT id FROM (
+        SELECT id, other_id, name FROM users
+        UNION ALL
+        SELECT id, other_id, name FROM users
+        ) u
+    ORDER BY name
+

--- a/server/src/main/java/io/crate/execution/dsl/projection/EvalProjection.java
+++ b/server/src/main/java/io/crate/execution/dsl/projection/EvalProjection.java
@@ -53,7 +53,8 @@ public class EvalProjection extends Projection {
     public static EvalProjection castValues(List<DataType<?>> targetTypes, List<Symbol> sources) {
         ArrayList<Symbol> casts = new ArrayList<>(targetTypes.size());
         boolean requiresCasts = false;
-        for (int i = 0; i < sources.size(); i++) {
+        assert sources.size() >= targetTypes.size() : "sources size must be >= targetTypes size";
+        for (int i = 0; i < targetTypes.size(); i++) {
             Symbol source = sources.get(i);
             DataType<?> targetType = targetTypes.get(i);
             InputColumn inputColumn = new InputColumn(i, source.valueType());

--- a/server/src/test/java/io/crate/execution/dsl/projection/EvalProjectionTest.java
+++ b/server/src/test/java/io/crate/execution/dsl/projection/EvalProjectionTest.java
@@ -30,7 +30,10 @@ import org.elasticsearch.test.ESTestCase;
 import org.junit.Test;
 
 import io.crate.expression.symbol.Literal;
+import io.crate.expression.symbol.Symbol;
 import io.crate.metadata.RowGranularity;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
 
 public class EvalProjectionTest extends ESTestCase {
 
@@ -44,5 +47,14 @@ public class EvalProjectionTest extends ESTestCase {
         var inProjection = new EvalProjection(in);
         assertThat(projection.requiredGranularity()).isEqualTo(inProjection.requiredGranularity());
         assertThat(projection.outputs()).isEqualTo(inProjection.outputs());
+    }
+
+    @Test
+    public void test_cast_target_types_can_be_less_than_source_outputs() {
+        List<DataType<?>> targetTypes = List.of(DataTypes.INTEGER);
+        List<Symbol> sources = List.of(Literal.of("1"), Literal.of("2"));
+        EvalProjection projection = EvalProjection.castValues(targetTypes, sources);
+        assertThat(projection).isNotNull();
+        assertThat(projection.outputs().size()).isEqualTo(targetTypes.size());
     }
 }

--- a/server/src/test/java/io/crate/planner/UnionPlannerTest.java
+++ b/server/src/test/java/io/crate/planner/UnionPlannerTest.java
@@ -279,4 +279,32 @@ public class UnionPlannerTest extends CrateDummyClusterServiceUnitTest {
         );
     }
 
+    /*
+     * https://github.com/crate/crate/issues/17341
+     */
+    @Test
+    public void test_union_with_order_by_on_non_selected_symbol() throws Exception {
+        String stmt = """
+            SELECT id FROM (
+                SELECT id, other_id, name FROM users
+                UNION ALL
+                SELECT id, other_id, name FROM users
+                ) u
+            ORDER BY name
+            """;
+        LogicalPlan logicalPlan = e.logicalPlan(stmt);
+        assertThat(logicalPlan).isEqualTo("""
+            Rename[id] AS u
+              └ Union[id]
+                ├ OrderBy[name ASC]
+                │  └ Collect[doc.users | [id, name] | true]
+                └ OrderBy[name ASC]
+                  └ Collect[doc.users | [id, name] | true]
+            """
+        );
+
+        // Ensure that this plan can be built without any exception
+        // (Ensure Eval can handle fewer outputs than its source)
+        e.plan(stmt);
+    }
 }

